### PR TITLE
Turbo fish assist: don't include lifetime parameters

### LINKED ISSUE #11219: "Add a turbofish" code action generates placeholders for lifetime arguments too, which is wrong
I requested that "add a turbofish" should respect the number of type arguments in  #11135 and the feature was immediately implemented in #11136 by @Vannevelj (thanks a lot!).  However, [it uses the number of "all" generic arguments](https://github.com/rust-analyzer/rust-analyzer/pull/11136/files#diff-2b33e63c85aeb932d56cdf7a2ef2b4ff6c7d2c69caa32b6fb992b68fd0b9d617R81) as the number of underscores, which may include lifetime parameters.   The code action should have actually checked if each argument is a type parameter, a const parameter, or a lifetime parameter, the last of which should not be included.

### PR DESCRIPTION
**Pull request by Vannevelj**
Fixes #11219

The issue talks about three different types of params: type, const & lifetime. I wasn't entirely sure which ones are intended to be included here so I've gone for the type & const params (i.e. exclude lifetime).

I've added a test case for both a lifetime param and a const param. I'm still making my way through the rust book (chapter 7, yay) so I'm not too sure yet what these are but my testing shows that this approach generates code that compiles.

### CODE REVIEW

File: crates/ide_assists/src/handlers/add_turbo_fish.rs [hunk 1]
```diff
@@ -77,7 +77,13 @@ pub(crate) fn add_turbo_fish(acc: &mut Assists, ctx: &AssistContext) -> Option<(
         }
     }
 
-    let number_of_arguments = generics.len();
+    let number_of_arguments = generics
+        .iter()
+        .filter(|param| match param {
+            hir::GenericParam::TypeParam(_) | hir::GenericParam::ConstParam(_) => true,
+            _ => false,
+        })
```
Comment (Veykril) [inline crates/ide_assists/src/handlers/add_turbo_fish.rs:85]:
```suggestion
        .filter(|param| matches!(param, hir::GenericParam::TypeParam(_) | hir::GenericParam::ConstParam(_)))
```
can be simplified a bit with a special macro exposed by the std prelude(it basically expands to the `match` you have written but is more concise)

File: crates/ide_assists/src/handlers/add_turbo_fish.rs [hunk 2]
```diff
@@ -339,4 +345,42 @@ fn main() {
...
+            r#"
+fn make<T, const N: usize>(t: T) {}
+fn main() {
+    make$0(3);
+}
+"#,
+            r#"
+fn make<T, const N: usize>(t: T) {}
+fn main() {
+    make::<${0:_,_}>(3);
```
Comment (Veykril) [inline crates/ide_assists/src/handlers/add_turbo_fish.rs:381]: Also, the snippet here(and on line `362`) are actually wrong(this is an oversight from your previous PR), this should be `make::<${1:_},${2:_}>(3);`

Comment (Veykril) [reply crates/ide_assists/src/handlers/add_turbo_fish.rs:381]: or rather `make::<${1:_},${0:_}>(3);`(since the order goes `1, 2, 3, ..., n - 1, n, 0`, basically 0 is always the last tabstop)

File: crates/ide_assists/src/handlers/add_turbo_fish.rs (hunk 1 above)
Comment (Vannevelj) [reply crates/ide_assists/src/handlers/add_turbo_fish.rs:85]: This is great, thanks! I was hoping there would be a less verbose way of writing this

File: crates/ide_assists/src/handlers/add_turbo_fish.rs (hunk 2 above)
Comment (Veykril) [reply crates/ide_assists/src/handlers/add_turbo_fish.rs:378]: Not saying you have to fix that in this PR btw, forgot to mention that. Just wanted to add that as a side note(though you are are welcome to fix this as well if you have the time)

bors d+

Comment (Vannevelj) [reply crates/ide_assists/src/handlers/add_turbo_fish.rs:378]: I don't mind fixing this now, I just need a little bit more information on what I'm doing here. What do the `1` and `0` represent in your example of `make::<${1:_},${0:_}>(3);`?

I presume I'd have to change it here: https://github.com/rust-analyzer/rust-analyzer/blob/2fb6f5e46a5f53a78c2f750dd3c8e79dd418c4c9/crates/ide_assists/src/handlers/add_turbo_fish.rs#L89
What about the `None` case underneath? Does that exclude the `1` and `0` bits?

Comment (Veykril) [reply crates/ide_assists/src/handlers/add_turbo_fish.rs:380]: So the $n things are tabstops, which define an order of where the cursor in the client can jump to(usually by pressing tab), `{$n:_}` is a special form of this that just defines a placeholder/template text(the `_` part is the placeholder text). The numbers define the order, where as 0 is always the last tab stop that gets jumpted to, otherwise 1 is the first, 2 is the second etc.

The None case below is for when the client doesnt support snippets(tabstops), so there we don't want any of the $ shenanigans.

So for the some branch we would want to see this `<${1:_},${2:_},${0:_}>` if we had 2 tabstops and for the None branch it should be `<_, _, _>`, if we had 3 generics as an example.

Comment (Vannevelj) [reply crates/ide_assists/src/handlers/add_turbo_fish.rs]: Thanks for the explanation! I've pushed a commit that sorts out these tabstops. The code is a bit wordy though so perhaps you might see an improvement I can make here?

It looks like the unit tests always test the `Some()` branch -- is there a way to turn snippet support off in the tests? That way I could write a test for the other case as well.

Comment (Veykril) [reply crates/ide_assists/src/handlers/add_turbo_fish.rs]: > It looks like the unit tests always test the Some() branch -- is there a way to turn snippet support off in the tests? That way I could write a test for the other case as well.

I don't think so, that is a good point that we should adress at some point though.

File: crates/ide_assists/src/handlers/add_turbo_fish.rs [hunk 3]
```diff
@@ -77,26 +77,46 @@ pub(crate) fn add_turbo_fish(acc: &mut Assists, ctx: &AssistContext) -> Option<(
...
         ident.text_range(),
         |builder| match ctx.config.snippet_cap {
             Some(cap) => {
-                let snip = format!("::<${{0:{}}}>", fish_head);
+                let snip = format!("::<{}>", get_snippet_fish_head(number_of_arguments));
                 builder.insert_snippet(cap, ident.text_range().end(), snip)
             }
             None => {
+                let fish_head =
+                    std::iter::repeat("_").take(number_of_arguments).collect::<Vec<_>>().join(", ");
```
Comment (Veykril) [inline crates/ide_assists/src/handlers/add_turbo_fish.rs:98]:
```suggestion
                    std::iter::repeat("_").take(number_of_arguments).format(", ");
```
(`Itertools` trait method)

File: crates/ide_assists/src/handlers/add_turbo_fish.rs [hunk 4]
```diff
@@ -77,26 +77,46 @@ pub(crate) fn add_turbo_fish(acc: &mut Assists, ctx: &AssistContext) -> Option<(
...
 }
 
+/// This will create a snippet string with tabstops marked
+fn get_snippet_fish_head(number_of_arguments: usize) -> String {
+    let mut fish_head = String::new();
+    let mut i = 1;
+    while i < number_of_arguments {
+        fish_head.push_str(&format!("${{{}:_}},", i));
+        i = i + 1;
+    }
```
Comment (Veykril) [inline crates/ide_assists/src/handlers/add_turbo_fish.rs:113]:
```suggestion
    let mut fish_head = String::new();
    for i in 1..number_of_arguments {
        fish_head.push_str(&format!("${{{}:_}},", i));
    }
```
which can be further simplified by using `Itertools`
```suggestion
    let mut fish_head = (1..number_of_arguments).format_with(", ", |i, f| {
        f(&format_args!("${{{}:_}},", i))
    });
```

Comment (Vannevelj) [reply crates/ide_assists/src/handlers/add_turbo_fish.rs]: This is great, thanks! I've applied your suggestions

**Comment by Veykril** [comment]
Thanks!
bors r+

**Comment by TonalidadeHidrica** [comment]
@Vannevelj Thank you again for the very fast implementation!  I can't believe you are new to Rust .  Amazing work.

**Comment by Vannevelj** [comment]
Thanks for the kind words!
